docs/requirements.md

```diff
@@ -0,0 +1,149 @@
+
+## Locale-sensitive layout
+
+Text layout is sensitive to locales, entailing a number of requirements. First is representation of locales. The underlying standard is [BCP 47]. We should try to use common Rust infrastructure for the parsing and representation. In addition, there are some utility functions that are likely specific to text, for example inferring script from other locale information (in ICU, this is done with the [`addLikelySubtags`] method).
+
+Locales manifest in two primary ways. First, they affect selection of fonts from a fallback list, especially Han unification. Second, they need to be plumbed to the shaping engine so OpenType [`locl`] features can be selected.
+
+Note that to really do this properly requires a locale *list,* not just a single locale as in some older interfaces. An example would be two users with en_US as their primary locale, but with ja_JP and zh_CN as their secondary locales. In this case, the secondary locale would be responsible for Han unification. In short, Han unification is controlled by the first CJK entry in the locale list. But Han unification is only the most visible example; in general, localization needs to be done from the first relevant entry, and the logic for what's relevant is complex and font-dependent.
+
+## Performance
+
+Text layout can be slow, and especially as more of the rendering pipeline is moving to the GPU, it can be the limiting factor. A large part of the design goal is performance, and here are some mechanisms to try to achieve that:
+
+* A layout cache. One subtlety is that it should be designed for multithreaded use, so a `Mutex<HashMap>` may replaced at some point with a concurrent hash map.
+
+* A fast representation of styles. Since style information can be expensive, both for storage and for hash-table lookup, we might design an arena, indexed by integer. Alternatively, we might use a small struct that covers the most common cases, with an optional `Box` for fancier style parameters such as OpenType features. In any case, locale is a large part of the puzzle.
+
+* Possibly fast paths for monospace fonts, including CJK. These have to be carefully designed to preserve semantics (unlike the "simple path" in Blink before it was removed).
+
+## Some open questions
+
+* Do shaping driven itemization, or use coverage data extracted from fonts (as in Minikin)?
+
+* Is layout at this level size-independent, or can it take into account pixel size dependent metrics (see [font-kit#34](https://github.com/pcwalton/font-kit/issues/34)).
```

Comment:
FWIW I’ve had bugs in [WeasyPrint](https://weasyprint.org/) because Pango was used in two places: during CSS layout with an abstract / vector cairo surface, then during painting with a cairo surface (potentially) backed by a pixel buffer. Pango would give different results, sometimes breaking lines at different places. I suspect this was due to font hinting (rounding the position of glyphs up to the next pixel boundary) affecting text layout based on device resolution (when there is one).

So, it would be nice if text layout could be resolution-independent, but I don’t know if that’s always possible depending on whether and how the glyph rendering part of the stack wants to do hinting.

Reply:
Right, I've seen very similar bugs when there isn't one source of truth. Even floating-point rounding errors can make measurement not fully repeatable.

The question of resolution independence is largely a question of whether you want to be forward-looking in rendering, where most displays are high-dpi, and the smoothness of transformation (like the pinch-to-zoom gesture) is more important than optimizing text for contrast, **or** whether you want to optimize for what I call the "GDI" world, where screens are low res, fonts are carefully hinted to render with good contrast on those screens, and you want layout to accommodate those fonts.

If we're using PathFinder for font rendering, it almost answers the question :) @pcwalton and I will discuss in person on Monday.